Fix StreamResultToBytes._write_packet()

haypo opened the pull request:
Call write() in a loop until all bytes are written. The write()
method doesn't ensure that all bytes are written.

This change should workaround the eventlet bug:
https://github.com/eventlet/eventlet/issues/248

haypo commented:
My change uses memoryview which isn't supported on Python 2.6. Does subunit support Python 2.6? If yes, something else should be used (buffer?).

haypo commented:
Oh, in Python 2, file.write(bytes) returns None. In fact, the code raises an IOError if the write is partial.

Maybe the loop in subunit should be made conditional: only for Python 3?

rbtcollins commented:
Yeah, so even in three file.write() returns None for the upper layers.

I think perhaps an isinstance check in the class constructor, and a decorator to do the loop?

haypo commented:
I added an if on the major python version.

haypo commented:
Oops, the Python 2 code was completly broken :-) It should now be fixed.

haypo commented:
Travis failed on Python 3.2 but it's unrelated to my change: hypothesis dependency failed to be installed because it doesn't support Python 3.2 (anymore?), it uses u'unicode' syntax which was removed in Python 3.0 (and reintroduced in Python 3.3).

haypo commented:
(Oops, I closed the pull request because I clicked on the wrong button...)

rbtcollins commented:
Could you please rebase this - I've fixed the 3.2 tests in master.

haypo commented:
> Could you please rebase this - I've fixed the 3.2 tests in master.

Ok, I rebased my change.

FYI my fix was merged into eventlet (https://github.com/eventlet/eventlet/commit/e3f88347d96f653d65f66a46b6bda618a6cb129d) but it was not released yet (https://github.com/eventlet/eventlet/issues/256).

rbtcollins commented:
This is released in 1.2. FWIW
